Wrappers/Python/cil/optimisation/algorithms/GD.py

```diff
@@ -19,137 +19,114 @@
 import numpy
 from cil.optimisation.algorithms import Algorithm
 import logging
+from cil.optimisation.utilities import ConstantStepSize, ArmijoStepSize
 
 log = logging.getLogger(__name__)
 
 
 class GD(Algorithm):
-    """Gradient Descent algorithm"""
-    def __init__(self, initial=None, objective_function=None, step_size=None, alpha=1e6, beta=0.5, rtol=1e-5, atol=1e-8, **kwargs):
+    """Gradient Descent algorithm
+
+    Parameters
+    ----------
+    initial: DataContainer (e.g. ImageData)
+        The initial point for the optimisation 
+    objective_function: CIL function with a defined gradient 
+        The function to be minimised. 
+    step_size: float, default = None 
+        Step size for gradient descent iteration if you want to use a constant step size. If left as None and do not pass a step_size_rule then the Armijio rule will be used to perform backtracking to choose a step size at each iteration. 
+    step_size_rule: class with a `get_step_size` method or a function that takes an initialised CIL function as an argument and outputs a step size, default is None
+        This could be a custom `step_size_rule` or one provided in :meth:`~cil.optimisation.utilities.StepSizeMethods`. If None is passed  then the algorithm will use either `ConstantStepSize` or `ArmijioStepSize` depending on if a `step_size` is provided. 
+    preconditioner: class with a `apply` method or a function that takes an initialised CIL function as an argument and modifies a provided `gradient`.
+            This could be a custom `preconditioner` or one provided in :meth:`~cil.optimisation.utilities.preconditoner`. If None is passed  then `self.gradient_update` will remain unmodified. 
+
+    rtol: positive float, default 1e-5
+        optional parameter defining the relative tolerance comparing the current objective function to 0, default 1e-5, see numpy.isclose
+    atol: positive float, default 1e-8
+        optional parameter defining the absolute tolerance comparing the current objective function to 0, default 1e-8, see numpy.isclose
+
+    """
+
+    def __init__(self, initial=None, objective_function=None, step_size=None, alpha=1e6, beta=0.5, rtol=1e-5, atol=1e-8, step_size_rule=None, preconditioner=None, **kwargs):
```

Comment:
Why do we need 4 parameters for a simple Armijo rule? Is this from eq(2) `https://projecteuclid.org/download/pdf_1/euclid.pjm/1102995080`?

- Is this tested with SIRF? I remembered there was a problem with GD and MR. Maybe not anymore. You can compare for example the scipy solution used in [e_advanced_recon](https://github.com/SyneRBI/SIRF-Exercises/blob/master/notebooks/MR/e_advanced_recon.ipynb)

```
from cil.optimisation.algorithms import GD
from cil.optimisation.utilities import ArmijoStepSize
from cil.optimisation.functions import LeastSquares
from cil.optimisation.utilities import MetricsDiagnostics, RSE

t = lambda x, y : np.sqrt(np.sum((np.abs(x - y))**2))
cb = MetricsDiagnostics(reference_image = img, verbose=1, metrics_dict={'rse':t}) 

f = LeastSquares(A = E, b=y, c=0.5)

gd = GD(initial=bwd_img, objective_function=f, step_size = 0.1,
        update_objective_interval=10, max_iteration=500)
gd.run(verbose=1, callback=[cb])

armj = ArmijoStepSize(initial=1., rho=0.5, c=0.5, iterations=50)
gd_armj = GD(initial=bwd_img, objective_function=f, step_size = armj,
        update_objective_interval=10, max_iteration=500)
gd_armj.run(verbose=1, callback=[cb])
```
<img width="1163" alt="Screenshot 2024-05-10 at 10 09 05" src="https://github.com/TomographicImaging/CIL/assets/22398586/eebe32a3-00fc-4c43-b8ef-f36683a557bc">

- It is strange that we have these arguments in the signature of GD and not in FISTA/ISTA.

- I think a for loop is better to terminate the line search based on fixed number of iterations. I am not sure which parameter `rtol`, `atol` can do this. Also, since we pass `step_size_rule` there is no need for `alpha=1e6, beta=0.5, rtol=1e-5, atol=1e-8` in the `__init__`. So the default behaviour of GD, can be as before but the `step_size_rule` can use these parameters.

Reply:
I think `rtol` and `atol` are stopping criterion for the GD algorithm and nothing to do with the Armijo rule. The parameters `alpha` and `beta` are for the armijo rule and I agree that since we pass `step_size_rule` there is no need for them in the `__init__` but I would like to keep backwards compatibility for now. I will add a deprecation warning if people pass `alpha` or `beta`.